openpgp unsupported feature

`hydroxide auth …` results in `openpgp: unsupported feature: public key algorithm 22`

I tried some of the suggestions:
If I replace all occurrences of `golang.org/x/crypto` in this repo with `github.com/ProtonMail/crypto`, `go build` fails with
```
build github.com/emersion/hydroxide/cmd/hydroxide: cannot load golang.org/x/crypto/openpgp/ecdh: cannot find module providing package golang.org/x/crypto/openpgp/ecdh
```
(also, `go get` puts `golang.org/x/crypto v0.0.0-20190325154230-a5d413f7728c // indirect` in `go.mod`)

If I replace `golang.org/x/crypto` with `github.com/keybase/go-crypto` instead, `go build` fails with
```
# github.com/emersion/hydroxide/protonmail
../../protonmail/contacts.go:149:53: too few values in &openpgp.Key literal
../../protonmail/crypto.go:48:68: too few values in openpgp.Key literal
../../protonmail/crypto.go:57:54: too few values in openpgp.Key literal
../../protonmail/crypto.go:81:68: too few values in openpgp.Key literal
../../protonmail/crypto.go:88:54: too few values in openpgp.Key literal
```

@nast90210 Tested nast90210/0.2.2 latest. It builds, but auth panics:
```
./hydroxide auth myusername
Password: 
2FA code: XXXXXX
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x298acc]

goroutine 1 [running]:
github.com/keybase/go-crypto/openpgp.checkDetachedSignature(0x0, 0x0, 0x4419e8, 0x252c180, 0x441ec8, 0x26c212c, 0x20, 0x397458, 0x252c101, 0x252c180)
        /home/thrall/go/pkg/mod/github.com/keybase/go-crypto@v0.0.0-20181127160227-255a5089e85a/openpgp/read.go:450 +0x148
github.com/keybase/go-crypto/openpgp.CheckDetachedSignature(0x0, 0x0, 0x4419e8, 0x252c180, 0x441ec8, 0x26c212c, 0x3e, 0x243cc00, 0xa7144)
        /home/thrall/go/pkg/mod/github.com/keybase/go-crypto@v0.0.0-20181127160227-255a5089e85a/openpgp/read.go:411 +0x44
github.com/emersion/hydroxide/protonmail.decodeModulus(0x288c2c0, 0x282, 0x0, 0x7cc08, 0x2470128, 0x287d000, 0x1000)
        /home/thrall/Dev/emersion/hydroxide/protonmail/srp.go:25 +0xf8
github.com/emersion/hydroxide/protonmail.srp(0x243cd64, 0x16, 0x20, 0x2496548, 0x16, 0x20, 0x39)
        /home/thrall/Dev/emersion/hydroxide/protonmail/srp.go:151 +0x28
github.com/emersion/hydroxide/protonmail.(*Client).Auth(0x243cf80, 0xbe8bf8f9, 0x9, 0x24849a0, 0x16, 0x276a380, 0x6, 0x2496548, 0x2486300, 0x208201, ...)
        /home/thrall/Dev/emersion/hydroxide/protonmail/auth.go:130 +0x60
main.main()
        /home/thrall/Dev/emersion/hydroxide/cmd/hydroxide/hydroxide.go:98 +0xaa4
```
(I'm tried a few times on the off-chance I mis-typed my password)

@nast90210 I've tested you latest version of the patch and got a  segmentation violation

```
 /t/hydroxide   0.2.2 ±  ./hydroxide auth ****
Password:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x723174]

goroutine 1 [running]:
github.com/keybase/go-crypto/openpgp.checkDetachedSignature(0x0, 0x0, 0x94d120, 0xc0004aa2d0, 0x94d7c0, 0xc0004d2058, 0x3e, 0xc0000d9b08, 0x3, 0xc000030070)
	/home/void/projects/go/pkg/mod/github.com/keybase/go-crypto@v0.0.0-20181127160227-255a5089e85a/openpgp/read.go:450 +0x1b4
github.com/keybase/go-crypto/openpgp.CheckDetachedSignature(...)
	/home/void/projects/go/pkg/mod/github.com/keybase/go-crypto@v0.0.0-20181127160227-255a5089e85a/openpgp/read.go:411
github.com/emersion/hydroxide/protonmail.decodeModulus(0xc0004ce2c0, 0x282, 0xc0000d9b18, 0xc0000d9b40, 0x66c166, 0xc000176210, 0x0)
	/tmp/hydroxide/protonmail/srp.go:25 +0x138
github.com/emersion/hydroxide/protonmail.srp(0xc0000d9c08, 0x8, 0x20, 0xc0000d22d0, 0x8, 0x20, 0xc00009efc0)
	/tmp/hydroxide/protonmail/srp.go:151 +0x4d
github.com/emersion/hydroxide/protonmail.(*Client).Auth(0xc0000d9f00, 0x7ffc6af7fba2, 0x4, 0xc0000b0370, 0x8, 0x0, 0x0, 0xc0000d22d0, 0x8, 0x8, ...)
	/tmp/hydroxide/protonmail/auth.go:130 +0x9f
main.main()
	/tmp/hydroxide/cmd/hydroxide/hydroxide.go:98 +0xdb8
```

need to rewrite some code with her custom crypto lib : 
`pkg/mod/github.com/emersion/hydroxide@v0.2.4/protonmail/contacts.go:167:66: not enough arguments in call to openpgp.CheckArmoredDetachedSignature
        have (openpgp.KeyRing, *bytes.Buffer, io.Reader)
        want (openpgp.KeyRing, io.Reader, io.Reader, *packet.Config)
pkg/mod/github.com/emersion/hydroxide@v0.2.4/protonmail/contacts.go:193:55: not enough arguments in call to openpgp.CheckArmoredDetachedSignature
        have (openpgp.KeyRing, *strings.Reader, *strings.Reader)
        want (openpgp.KeyRing, io.Reader, io.Reader, *packet.Config)
pkg/mod/github.com/emersion/hydroxide@v0.2.4/protonmail/crypto.go:39:15: subkey.Sig.KeyExpired undefined (type *packet.Signature has no field or method KeyExpired)
pkg/mod/github.com/emersion/hydroxide@v0.2.4/protonmail/crypto.go:56:137: i.SelfSignature.KeyExpired undefined (type *packet.Signature has no field or method KeyExpired)
pkg/mod/github.com/emersion/hydroxide@v0.2.4/protonmail/crypto.go:73:15: subkey.Sig.KeyExpired undefined (type *packet.Signature has no field or method KeyExpired)
pkg/mod/github.com/emersion/hydroxide@v0.2.4/protonmail/crypto.go:87:80: i.SelfSignature.KeyExpired undefined (type *packet.Signature has no field or method KeyExpired)
pkg/mod/github.com/emersion/hydroxide@v0.2.4/protonmail/srp.go:28:42: not enough arguments in call to openpgp.CheckDetachedSignature
        have (nil, *bytes.Reader, io.Reader)
        want (openpgp.KeyRing, io.Reader, io.Reader, *packet.Config)`

--- a/go.mod
+++ b/go.mod
@@ -19,3 +19,5 @@ require (
 	golang.org/x/sys v0.0.0-20200223170610-d5e6a3e2c0ae // indirect
 	gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127 // indirect
 )
+
+replace golang.org/x/crypto => github.com/ProtonMail/crypto v0.0.0-20200605105621-11f6ee2dd602

--- a/protonmail/contacts.go
+++ b/protonmail/contacts.go
@@ -164,7 +164,7 @@ func (r *detachedSignatureReader) Read(p []byte) (n int, err error) {
 	n, err = r.body.Read(p)
 	if err == io.EOF && !r.eof {
 		// Check signature
-		signer, signatureError := openpgp.CheckArmoredDetachedSignature(r.keyring, &r.signed, r.signature)
+		signer, signatureError := openpgp.CheckArmoredDetachedSignature(r.keyring, &r.signed, r.signature, nil)
 		r.md.IsSigned = true
 		r.md.SignatureError = signatureError
 		if signer != nil {
@@ -190,7 +190,7 @@ func (card *ContactCard) Read(keyring openpgp.KeyRing) (*openpgp.MessageDetails,
 
 		signed := strings.NewReader(card.Data)
 		signature := strings.NewReader(card.Signature)
-		signer, err := openpgp.CheckArmoredDetachedSignature(keyring, signed, signature)
+		signer, err := openpgp.CheckArmoredDetachedSignature(keyring, signed, signature, nil)
 		md.IsSigned = true
 		md.SignatureError = err
 		if signer != nil {

--- a/protonmail/crypto.go
+++ b/protonmail/crypto.go
@@ -36,7 +36,7 @@ func encryptionKey(e *openpgp.Entity, now time.Time) (openpgp.Key, bool) {
 		if subkey.Sig.FlagsValid &&
 			subkey.Sig.FlagEncryptCommunications &&
 			subkey.PublicKey.PubKeyAlgo.CanEncrypt() &&
-			!subkey.Sig.KeyExpired(now) &&
+			!subkey.PublicKey.KeyExpired(subkey.Sig, now) &&
 			(maxTime.IsZero() || subkey.Sig.CreationTime.After(maxTime)) {
 			candidateSubkey = i
 			maxTime = subkey.Sig.CreationTime
@@ -53,7 +53,7 @@ func encryptionKey(e *openpgp.Entity, now time.Time) (openpgp.Key, bool) {
 	// assume that the primary key is ok. Or, if the primary key is
 	// marked as ok to encrypt to, then we can obviously use it.
 	i := primaryIdentity(e)
-	if !i.SelfSignature.FlagsValid || i.SelfSignature.FlagEncryptCommunications && e.PrimaryKey.PubKeyAlgo.CanEncrypt() && !i.SelfSignature.KeyExpired(now) {
+	if !i.SelfSignature.FlagsValid || i.SelfSignature.FlagEncryptCommunications && e.PrimaryKey.PubKeyAlgo.CanEncrypt() && !i.SelfSignature.SigExpired(now) {
 		return openpgp.Key{e, e.PrimaryKey, e.PrivateKey, i.SelfSignature}, true
 	}
 
@@ -70,7 +70,7 @@ func signingKey(e *openpgp.Entity, now time.Time) (openpgp.Key, bool) {
 		if subkey.Sig.FlagsValid &&
 			subkey.Sig.FlagSign &&
 			subkey.PublicKey.PubKeyAlgo.CanSign() &&
-			!subkey.Sig.KeyExpired(now) {
+			!subkey.PublicKey.KeyExpired(subkey.Sig, now) {
 			candidateSubkey = i
 			break
 		}
@@ -84,7 +84,7 @@ func signingKey(e *openpgp.Entity, now time.Time) (openpgp.Key, bool) {
 	// If we have no candidate subkey then we assume that it's ok to sign
 	// with the primary key.
 	i := primaryIdentity(e)
-	if !i.SelfSignature.FlagsValid || i.SelfSignature.FlagSign && !i.SelfSignature.KeyExpired(now) {
+	if !i.SelfSignature.FlagsValid || i.SelfSignature.FlagSign && !i.SelfSignature.SigExpired(now) {
 		return openpgp.Key{e, e.PrimaryKey, e.PrivateKey, i.SelfSignature}, true
 	}
 

--- a/protonmail/srp.go
+++ b/protonmail/srp.go
@@ -8,12 +8,10 @@ import (
 	"errors"
 	"fmt"
 	"io"
-	"log"
 	"math/big"
 
 	"golang.org/x/crypto/openpgp"
 	"golang.org/x/crypto/openpgp/clearsign"
-	openpgperrors "golang.org/x/crypto/openpgp/errors"
 )
 
 var randReader io.Reader = rand.Reader
@@ -45,9 +43,8 @@ func decodeModulus(msg string) ([]byte, error) {
 	}
 
 	_, err = openpgp.CheckDetachedSignature(modulusKeyring, bytes.NewReader(block.Bytes), block.ArmoredSignature.Body, nil)
-	if err != nil && err != openpgperrors.ErrUnknownIssuer {
-		//return nil, fmt.Errorf("failed to decode modulus: %v", err)
-		log.Println("warning: failed to check SRP modulus signature:", err)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check modulus signature: %v", err)
 	}
 
 	b, err := base64.StdEncoding.DecodeString(string(block.Plaintext))

--- a/protonmail/srp.go
+++ b/protonmail/srp.go
@@ -1,19 +1,17 @@
 package protonmail
 
 import (
-	"bytes"
 	"crypto/rand"
 	"crypto/subtle"
 	"encoding/base64"
 	"errors"
 	"fmt"
 	"io"
-	"log"
+	// "log"
 	"math/big"
 
-	"golang.org/x/crypto/openpgp"
 	"golang.org/x/crypto/openpgp/clearsign"
-	openpgperrors "golang.org/x/crypto/openpgp/errors"
+	// openpgperrors "golang.org/x/crypto/openpgp/errors"
 )
 
 var randReader io.Reader = rand.Reader
@@ -25,11 +23,12 @@ func decodeModulus(msg string) ([]byte, error) {
 	}
 
 	// TODO: check signature and signature key
-	_, err := openpgp.CheckDetachedSignature(nil, bytes.NewReader(block.Plaintext), block.ArmoredSignature.Body)
-	if err != nil && err != openpgperrors.ErrUnknownIssuer {
+	// FIXME: segfaults
+	// _, err := block.VerifySignature(nil, nil)
+	// if err != nil && err != openpgperrors.ErrUnknownIssuer {
 		//return nil, fmt.Errorf("failed to decode modulus: %v", err)
-		log.Println("warning: failed to check SRP modulus signature:", err)
-	}
+	//	log.Println("warning: failed to check SRP modulus signature:", err)
+	//}
 
 	b, err := base64.StdEncoding.DecodeString(string(block.Plaintext))
 	if err != nil {

--- a/protonmail/srp.go
+++ b/protonmail/srp.go
@@ -1,20 +1,37 @@
 package protonmail
 
 import (
+	"bytes"
 	"crypto/rand"
 	"crypto/subtle"
 	"encoding/base64"
 	"errors"
 	"fmt"
 	"io"
-	// "log"
+	"log"
 	"math/big"
 
+	"golang.org/x/crypto/openpgp"
 	"golang.org/x/crypto/openpgp/clearsign"
-	// openpgperrors "golang.org/x/crypto/openpgp/errors"
+	openpgperrors "golang.org/x/crypto/openpgp/errors"
 )
 
 var randReader io.Reader = rand.Reader
+
+// Public key for SRP verification
+// From https://github.com/ProtonMail/proton-bridge/blob/99721b6577fe9079ac7547f11fc77e5090cdd31b/pkg/srp/srp.go#L41-L52
+const modulusPubkey = `-----BEGIN PGP PUBLIC KEY BLOCK-----
+
+xjMEXAHLgxYJKwYBBAHaRw8BAQdAFurWXXwjTemqjD7CXjXVyKf0of7n9Ctm
+L8v9enkzggHNEnByb3RvbkBzcnAubW9kdWx1c8J3BBAWCgApBQJcAcuDBgsJ
+BwgDAgkQNQWFxOlRjyYEFQgKAgMWAgECGQECGwMCHgEAAPGRAP9sauJsW12U
+MnTQUZpsbJb53d0Wv55mZIIiJL2XulpWPQD/V6NglBd96lZKBmInSXX/kXat
+Sv+y0io+LR8i2+jV+AbOOARcAcuDEgorBgEEAZdVAQUBAQdAeJHUz1c9+KfE
+kSIgcBRE3WuXC4oj5a2/U3oASExGDW4DAQgHwmEEGBYIABMFAlwBy4MJEDUF
+hcTpUY8mAhsMAAD/XQD8DxNI6E78meodQI+wLsrKLeHn32iLvUqJbVDhfWSU
+WO4BAMcm1u02t4VKw++ttECPt+HUgPUq5pqQWe5Q2cW4TMsE
+=Y4Mw
+-----END PGP PUBLIC KEY BLOCK-----`
 
 func decodeModulus(msg string) ([]byte, error) {
 	block, _ := clearsign.Decode([]byte(msg))
@@ -22,13 +39,16 @@ func decodeModulus(msg string) ([]byte, error) {
 		return nil, errors.New("invalid SRP modulus signed PGP block")
 	}
 
-	// TODO: check signature and signature key
-	// FIXME: segfaults
-	// _, err := block.VerifySignature(nil, nil)
-	// if err != nil && err != openpgperrors.ErrUnknownIssuer {
+	modulusKeyring, err := openpgp.ReadArmoredKeyRing(bytes.NewReader([]byte(modulusPubkey)))
+	if err != nil {
+		return nil, fmt.Errorf("cannot read modulus pubkey: %v", err)
+	}
+
+	_, err = openpgp.CheckDetachedSignature(modulusKeyring, bytes.NewReader(block.Bytes), block.ArmoredSignature.Body, nil)
+	if err != nil && err != openpgperrors.ErrUnknownIssuer {
 		//return nil, fmt.Errorf("failed to decode modulus: %v", err)
-	//	log.Println("warning: failed to check SRP modulus signature:", err)
-	//}
+		log.Println("warning: failed to check SRP modulus signature:", err)
+	}
 
 	b, err := base64.StdEncoding.DecodeString(string(block.Plaintext))
 	if err != nil {

